### PR TITLE
Add hexagon implementations for compiler_rt builtins

### PR DESCRIPTION
**Pull request by androm3da**

### CODE REVIEW

File: lib/compiler_rt/hexagon.zig [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+    );
+}
+
+comptime {
+    if (builtin.cpu.arch == .hexagon) {
+        @export(__hexagon_divsi3, .{ .name = "__hexagon_divsi3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_udivmoddi4, .{ .name = "__hexagon_udivmoddi4", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_umodsi3, .{ .name = "__hexagon_umodsi3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtf, .{ .name = "__hexagon_sqrtf", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_udivmodsi4, .{ .name = "__hexagon_udivmodsi4", .linkage = common.linkage, .visibility = common.visibility });
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:601]: Is this one actually used by LLVM?

**Comment by alexrp** [comment]
I see these also referenced in LLVM:

* `__hexagon_floattidf`
* `__hexagon_floattisf`
* `__hexagon_fixunssfti`
* `__hexagon_fixunsdfti`
* `__hexagon_fixsfti`
* `__hexagon_fixdfti`
* `__hexagon_fast2_sqrtf`
* `__hexagon_fast2_sqrtdf2`
* `__hexagon_adddf3` and `__hexagon_fast_adddf3`
* `__hexagon_subdf3` and `__hexagon_fast_subdf3`
* `__hexagon_muldf3` and `__hexagon_fast_muldf3`
* `__hexagon_divdf3` and `__hexagon_fast_divdf3`
* `__hexagon_divsf3` and `__hexagon_fast_divsf3`

Do we need these?

File: lib/compiler_rt/hexagon.zig (hunk 1 above)
Comment (androm3da) [reply lib/compiler_rt/hexagon.zig:601]: No, I don't believe so.  I included it in order to preserve the published builtins from `llvm-project/compiler-rt` but I'm happy to omit it.

**Comment by androm3da** [comment]
> I see these also referenced in LLVM:
> 
> * `__hexagon_floattidf`
> * `__hexagon_floattisf`
> * `__hexagon_fixunssfti`
> * `__hexagon_fixunsdfti`
> * `__hexagon_fixsfti`
> * `__hexagon_fixdfti`

These are not defined in `compiler-rt/lib/builtins/hexagon/` so I think they're safe to omit.  But I'll follow up with the team to understand why these references aren't satisfied.

> * `__hexagon_fast2_sqrtf`
> * `__hexagon_fast2_sqrtdf2`
> * `__hexagon_adddf3` and `__hexagon_fast_adddf3`
> * `__hexagon_subdf3` and `__hexagon_fast_subdf3`
> * `__hexagon_muldf3` and `__hexagon_fast_muldf3`
> * `__hexagon_divdf3` and `__hexagon_fast_divdf3`
> * `__hexagon_divsf3` and `__hexagon_fast_divsf3`
> 
> Do we need these?

These were unintentionally omitted.  I'll see about revising this PR to include these.

I assume it's safe/permitted/expected for me to use `.set` directives to create aliases from within an inline assembly block?  Otherwise I will need to find another mechanism to publish the aliases.

**Comment by alexrp** [comment]
> I assume it's safe/permitted/expected for me to use `.set` directives to create aliases from within an inline assembly block? Otherwise I will need to find another mechanism to publish the aliases.

Can you clarify what aliases you're referring to?

File: lib/compiler_rt/hexagon.zig (hunk 1 above)
Comment (alexrp) [reply lib/compiler_rt/hexagon.zig:601]: I think we can just omit it then. We only really care about those that the compiler actually uses in practice.

**Comment by androm3da** [comment]
> Can you clarify what aliases you're referring to?

The `__hexagon_fast_*` / `__qdsp_*` / `__hexagon_fast2_*` aliases, for example:

https://github.com/llvm/llvm-project/blob/e14827f0828d14ef17ab76316e8449d1b76e2617/compiler-rt/lib/builtins/hexagon/dfaddsub.S#L58-L71

https://github.com/rust-lang/compiler-builtins/blob/a09218f1c4d23ffbd97d68f0fefb5feed2469dc5/src/hexagon/dfaddsub.s#L7-L15

**Comment by androm3da** [comment]
> Hmm, I'm not sure if doing that in inline assembly is a good idea. 🤔 But if these are just aliases of the same symbol, perhaps you can just do multiple `@export()`s with different names?

I thought such a thing might be possible, so yeah, that'll do.  I will work on it.

File: lib/compiler_rt/hexagon.zig [hunk 2]
```diff
@@ -0,0 +1,1852 @@
...
+        \\   r1:0 = convert_sf2df(r28)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+comptime {
+    if (builtin.cpu.arch == .hexagon) {
+        @export(__hexagon_adddf3, .{ .name = "__hexagon_adddf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_adddf3, .{ .name = "__qdsp_adddf3", .linkage = common.linkage, .visibility = common.visibility });
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:1812]: From what I can see, LLVM doesn't use the `__qdsp_`-prefixed names. Perhaps this was for an older compiler?

File: lib/compiler_rt/hexagon.zig [hunk 3]
```diff
@@ -0,0 +1,1852 @@
...
+        @export(__hexagon_adddf3, .{ .name = "__hexagon_fast_adddf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_adddf3, .{ .name = "__hexagon_fast2_adddf3", .linkage = common.linkage, .visibility = common.visibility });
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:1814]: For the `__hexagon_fast2_` prefix, it seems like LLVM only uses `__hexagon_fast2_sqrtf` and `__hexagon_fast2_sqrtdf2`.

File: lib/compiler_rt/hexagon.zig [hunk 4]
```diff
@@ -0,0 +1,1852 @@
...
+        @export(__hexagon_muldf3, .{ .name = "__hexagon_fast_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_muldf3, .{ .name = "__hexagon_fast2_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__qdsp_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast2_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_sqrt", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__qdsp_sqrt", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast_sqrt", .linkage = common.linkage, .visibility = common.visibility });
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:1833]: Ditto for this one.

File: lib/compiler_rt/hexagon.zig (hunk 3 above)
Comment (alexrp) [reply lib/compiler_rt/hexagon.zig:1814]: Is there code in the wild that depends on them existing in some sort of "RTABI" capacity (a la Arm)? If that's the case, that'd be a strong argument in favor of keeping them all. Other than that, we only really care about the exact set needed by the compiler.

Comment (androm3da) [reply lib/compiler_rt/hexagon.zig]: None that I know of.  I'll remove unused ones.

File: lib/compiler_rt/hexagon.zig [hunk 5]
```diff
@@ -0,0 +1,1852 @@
...
+        @export(__hexagon_divdf3, .{ .name = "__hexagon_fast_divdf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divdf3, .{ .name = "__hexagon_fast2_divdf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_muldf3, .{ .name = "__hexagon_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_muldf3, .{ .name = "__qdsp_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_muldf3, .{ .name = "__hexagon_fast_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_muldf3, .{ .name = "__hexagon_fast2_muldf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__qdsp_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast_sqrtdf2", .linkage = common.linkage, .visibility = common.visibility });
```
[the remark this answers is not in the document]
Comment (androm3da) [reply lib/compiler_rt/hexagon.zig]: fixed

File: lib/compiler_rt/hexagon.zig (hunk 4 above)
Comment (androm3da) [reply lib/compiler_rt/hexagon.zig]: fixed

File: lib/compiler_rt/hexagon.zig [hunk 6]
```diff
@@ -0,0 +1,1852 @@
...
+        @export(__hexagon_sqrtdf2, .{ .name = "__qdsp_sqrt", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast_sqrt", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtdf2, .{ .name = "__hexagon_fast2_sqrt", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divsf3, .{ .name = "__hexagon_divsf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divsf3, .{ .name = "__qdsp_divsf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divsf3, .{ .name = "__hexagon_fast_divsf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divsf3, .{ .name = "__hexagon_fast2_divsf3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_divsi3, .{ .name = "__hexagon_divsi3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_udivmoddi4, .{ .name = "__hexagon_udivmoddi4", .linkage = common.linkage, .visibility = common.visibility });
```
[the remark this answers is not in the document]
Comment (androm3da) [reply lib/compiler_rt/hexagon.zig]: fixed

**Comment by androm3da** [comment]
Addressed review comments, fixed `__hexagon_subdf3` issue (thanks @sidmanning)

**Review by SidManning — APPROVED** [review]

File: lib/compiler_rt/hexagon.zig [hunk 7]
```diff
@@ -0,0 +1,1788 @@
...
+        \\  }:endloop0
+        \\  {
+        \\   p0 = cmp.gtu(r2,r1)
+        \\   if (!p0.new) r0 = add(r0,r3)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+fn __hexagon_adddf3() align(1 << 5) callconv(.naked) noreturn {
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:510]:
```suggestion
fn __hexagon_adddf3() align(32) callconv(.naked) noreturn {
```

File: lib/compiler_rt/hexagon.zig [hunk 8]
```diff
@@ -0,0 +1,1788 @@
...
+fn __hexagon_subdf3() align(32) callconv(.naked) noreturn {
+    asm volatile (
+        \\ {
+        \\   r3 = togglebit(r3,#31)
+        \\   jump ##__hexagon_adddf3
+        \\ }
+    );
+}
+
+fn __hexagon_divdf3() align(1 << 5) callconv(.naked) noreturn {
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:822]:
```suggestion
fn __hexagon_divdf3() align(32) callconv(.naked) noreturn {
```

File: lib/compiler_rt/hexagon.zig [hunk 9]
```diff
@@ -0,0 +1,1788 @@
...
+        \\   r10 = ##0x7f800001
+        \\  }
+        \\  {
+        \\   r1:0 = convert_sf2df(r10)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+fn __hexagon_muldf3() align(1 << 5) callconv(.naked) noreturn {
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:1190]:
```suggestion
fn __hexagon_muldf3() align(32) callconv(.naked) noreturn {
```

File: lib/compiler_rt/hexagon.zig [hunk 10]
```diff
@@ -0,0 +1,1788 @@
...
+        \\   r3 = extract(r3,#1,#31)
+        \\  }
+        \\  {
+        \\   r1 ^= asl(r3,#31)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+fn __hexagon_sqrtdf2() align(1 << 5) callconv(.naked) noreturn {
```
Comment (alexrp) [inline lib/compiler_rt/hexagon.zig:1495]:
```suggestion
fn __hexagon_sqrtdf2() align(32) callconv(.naked) noreturn {
```

**Comment by alexrp** [comment]
> These are not defined in `compiler-rt/lib/builtins/hexagon/` so I think they're safe to omit. But I'll follow up with the team to understand why these references aren't satisfied.

This looks like an LLVM bug to me. They seem to be used for `i128` <-> `f32`/`f64` conversions. Clang disables `__int128` for Hexagon, so it's not an issue there, but it is an issue for Zig:

```console
❯ cat test.zig
export fn cvt(d: i128) f64 {
    return @floatFromInt(d);
}
❯ zig4 build-lib test.zig -fno-compiler-rt -target hexagon-freestanding
❯ llvm-objdump -r --disassemble-symbols=cvt libtest.a

libtest.a(./libtest.a.o):       file format elf32-hexagon

Disassembly of section .text:

00000000 <cvt>:
       0:       02 c0 9d a0     a09dc002 {      allocframe(#0x10) }
       4:       04 c2 03 f5     f503c204 {      r5:4 = combine(r3,r2) }
       8:       04 c0 01 f5     f501c004 {      r5:4 = combine(r1,r0) }
       c:       ff e2 de a7     a7dee2ff {      memd(r30+#-0x8) = r3:2 }
      10:       fe e0 de a7     a7dee0fe {      memd(r30+#-0x10) = r1:0 }
      14:       00 c0 00 5a     5a00c000 {      call 0x14 }                     00000014:  R_HEX_B22_PCREL      __hexagon_floattidf
      18:       1e c0 1e 96     961ec01e {      dealloc_return }
```

**Comment by alexrp** [comment]
So, I think [these calls](https://github.com/llvm/llvm-project/blob/aa5dc539e91824cbc224214e69d62d46db21af6e/llvm/lib/Target/Hexagon/HexagonISelLowering.cpp#L1863-L1868) should just all be deleted. That way, the code generator will emit calls to the usual `__floattidf`, etc. (Unless of course these Hexagon-optimized routines actually exist somewhere and just haven't been upstreamed to LLVM yet.)

**Comment by androm3da** [comment]
> (Unless of course these Hexagon-optimized routines actually exist somewhere and just haven't been upstreamed to LLVM yet.)

I think this may be the case.  I'll check and update the PR soon.

**Comment by alexrp** [comment]
Ok, given https://github.com/llvm/llvm-project/pull/117423, I think this PR is good to go after the remaining `align(32)` suggestions are accepted.

**Comment by androm3da** [comment]
> Ok, given [llvm/llvm-project#117423](https://github.com/llvm/llvm-project/pull/117423), I think this PR is good to go after the remaining `align(32)` suggestions are accepted.

Ok - I've taken those, thanks.

re: llvm/llvm-project#117423 -- if it turns out that it's not preferred to remove those, then I'll probably add the specializations to `compiler-rt` instead (and then corresponding ones here).

**Pull request merged**